internal/core/worker/universal/worker.go

```diff
@@ -0,0 +1,139 @@
+
+	"github.com/freeverseio/laos-universal-node/internal/config"
+	"github.com/freeverseio/laos-universal-node/internal/core/processor/universal"
+	contractDiscoverer "github.com/freeverseio/laos-universal-node/internal/core/processor/universal/discoverer"
+	contractUpdater "github.com/freeverseio/laos-universal-node/internal/core/processor/universal/updater"
+	"github.com/freeverseio/laos-universal-node/internal/platform/scan"
+	"github.com/freeverseio/laos-universal-node/internal/platform/state"
+)
+
+type Worker interface {
+	Run(ctx context.Context) error
+}
+
+type worker struct {
+	waitingTime time.Duration
+	processor   universal.Processor
+}
+
+func NewWorker(c *config.Config,
+	client scan.EthClient,
+	scanner scan.Scanner,
+	stateService state.Service,
+	discoverer contractDiscoverer.Discoverer,
+	updater contractUpdater.Updater,
+) Worker {
+	return &worker{
+		waitingTime: c.WaitingTime,
+		processor: universal.NewProcessor(
+			client,
+			stateService,
+			scanner,
+			c,
+			discoverer,
+			updater),
+	}
+}
+
+func (w *worker) Run(ctx context.Context) error {
+	slog.Info("starting universal worker")
+	startingBlock, err := w.processor.GetInitStartingBlock(ctx)
+	if err != nil {
+		return err
+	}
+
+	evoSynced := true
+	lastBlock := startingBlock
+	for {
+		select {
+		case <-ctx.Done():
+			slog.Info("context canceled")
+			return nil
+		default:
+			slog.Debug("executing block range", "startingBlock", startingBlock, "lastBlock", lastBlock, "evoSynced", evoSynced)
+			prevLastBlock, wasEvoSynced, err := w.executeUniversalBlockRange(ctx, evoSynced, startingBlock, lastBlock)
+			if err != nil {
+				var reorgErr universal.ReorgError
+				if errors.As(err, &reorgErr) {
+					slog.Error("ownership chain reorganization detected",
+						"blockNumber", reorgErr.Block,
+						"chainHash", reorgErr.ChainHash.String(),
+						"storageHash", reorgErr.StorageHash.String())
+					slog.Info("***************************************************************************************")
+					slog.Info("Please wipe out the database before running the node again.")
+					slog.Info("***************************************************************************************")
+					return reorgErr
+				}
+				break
+			}
+
+			evoSynced = wasEvoSynced
+			// if evo is not synced (evoSynced == false) the lastBlock should be the one that is previously and don't update startingBlock
+			// otherwise update startingBlock to the lastBlock + 1 and the new lastBlock will be calculated in the function
+			if !evoSynced {
+				lastBlock = prevLastBlock
+				break
+			}
+			startingBlock = prevLastBlock + 1
+		}
+	}
+}
+
+func (w *worker) executeUniversalBlockRange(ctx context.Context,
+	evoSynced bool,
+	startingBlock,
+	lastBlock uint64,
+) (previousLastBlock uint64, wasEvoSynced bool, err error) {
+	if evoSynced {
+		lastBlock, err = w.processor.GetLastBlock(ctx, startingBlock)
+		if err != nil {
+			slog.Debug("Error obtaining the last block", "error", err)
```

Comment:
We don't need this as the error is logged by `GetLastBlock``

Reply:
ok